src/client/pydaos/__init__.py

```diff
@@ -25,42 +25,69 @@
 import sys
 import atexit
 
+dc = None
+
 DAOS_MAGIC = 0x7A89
 
 # pylint: disable=import-error
 if sys.version_info < (3, 0):
-    import pydaos_shim_27 as pydaos_shim
+    from . import pydaos_shim_27 as pydaos_shim
 else:
-    import pydaos_shim_3 as pydaos_shim
+    from . import pydaos_shim_3 as pydaos_shim
 # pylint: enable=import-error
 
-from pydaos_core import *
-
-__all__ = ["pydaos_core"]
-
+# Define the PyDError class here before doing the pydaos_core import so that
+# it's accessible from within the module.
 class PyDError(Exception):
-    """
-    PyDAOS exception when operation cannot be completed.
-    DER_* error code is printed in both integer and string format.
-    """
+    """PyDAOS exception when operation cannot be completed."""
+
+    # DER_* error code is printed in both integer and string format where
+    # possible.  There is an odd effect with daos_init() errors that
+    # pydaos_shim is valid during __init__ but None during __str__ so format
+    # the string early and just report it later on.
     def __init__(self, message, rc):
-        self.message = message
-        self.rc = rc
+        err = pydaos_shim.err_to_str(DAOS_MAGIC, rc)
+        if err:
+            self.message = '{}: {}'.format(message, err)
+        else:
+            self.message = '{}: {}'.format(message, rc)
 
     def __str__(self):
-        err = pydaos_shim.err_to_str(DAOS_MAGIC, self.rc)
-        if err is not None:
-            return self.message + ": " + err
-        else:
-            return self.message
+        return self.message
 
-# Initialize DAOS
-_rc = pydaos_shim.daos_init(DAOS_MAGIC)
-if _rc != pydaos_shim.DER_SUCCESS:
-    raise PyDError("Failed to initialize DAOS", _rc)
+class DaosClient():
+    """DaosClient object"""
+
+    # Created automatically as module is imported, and the
+    # local ref is dropped when it's unloaded.
+    def __init__(self):
+        # Initialize DAOS
+        self.connected = False
+        _rc = pydaos_shim.daos_init(DAOS_MAGIC)
+        if _rc != pydaos_shim.DER_SUCCESS:
+            raise PyDError("Failed to initialize DAOS", _rc)
+        self.connected = True
+
+    def _close(self):
+        if not self.connected:
+            return
+        rc = pydaos_shim.daos_fini(DAOS_MAGIC)
+        if rc != pydaos_shim.DER_SUCCESS:
+            raise PyDError("Failed to cleanup DAOS", rc)
+        self.connected = False
+
+    def __del__(self):
+        if not pydaos_shim or not self.connected:
+            return
+        self._close()
+
+dc = DaosClient()
 
 @atexit.register
 def _cleanup():
-    rc = pydaos_shim.daos_fini(DAOS_MAGIC)
-    if rc != pydaos_shim.DER_SUCCESS:
-        raise PyDError("Failed to cleanup DAOS", rc)
+    global dc
+    dc = None;
```

Comment:
not sure why we need to rely on the garbage collector for that instead of calling daos_fini() as in the original version. What does the DaosClient() buys us here?

Reply:
I was using references to prevent things being deleted so it seemed natural to use reference counting to delete things as well, although it would be clearer if it was explicit.

I could probably remove DaosClient() again as it didn't work as well as I'd hoped, although i have a strong preference for removing the automatic calling of daos_init() at import anyway - I was hoping to do that in a further patch.